source/common/http/http1/codec_impl.cc

```diff
@@ -411,8 +411,9 @@ int ServerConnectionImpl::onHeadersComplete(HeaderMapImplPtr&& headers) {
         http_parser_pause(&parser_, 1);
       }
 
-    } else if (parser_.method == http_method::HTTP_POST ||
-               parser_.method == http_method::HTTP_PUT) {
+    } else if ((parser_.method == http_method::HTTP_POST ||
+                parser_.method == http_method::HTTP_PUT) &&
+               headers->ContentLength() == nullptr) {
```

Comment:
It used to be in RFC 2616 that the message body for requests dependent upon the method being sent. The bis standards loosened the requirements so that the framing layer could be decoupled from the method semantics.

Requiring a Content-Length header for POST / PUT for HTTP/1.1 should be fine; however, it may break some hand-written clients that had not been setting content-length and letting the server assume an empty entity body.

IIRC there is some security considerations to requiring Content-Length as a mitigation for HTTP Request Smuggling attacks. This assumes that the server is also correctly failing requests with multiple content-length headers that do not match.